src/test/spec/unified_runner/matcher.rs

```diff
@@ -0,0 +1,235 @@
+                return false;
+            }
+
+            for (actual, expected) in actual_array.iter().zip(expected_array) {
+                if !results_match(Some(actual), expected) {
+                    return false;
+                }
+            }
+
+            true
+        }
+        _ => match actual {
+            Some(actual) => actual == expected,
+            None => false,
+        },
+    }
+}
+
+fn special_operator_matches(special_op: (&String, &Bson), actual: Option<&Bson>) -> bool {
+    match special_op.0.as_ref() {
+        "$$exists" => special_op.1.as_bool().unwrap() == actual.is_some(),
+        "$$type" => type_matches(special_op.1, actual.unwrap()),
+        "$$unsetOrMatches" => {
+            if let Some(bson) = actual {
+                results_match(Some(special_op.1), bson)
+            } else {
+                true
+            }
+        }
+        "$$sessionLsid" => panic!("Explicit sessions not implemented"),
+        other => panic!("unknown special operator: {}", other),
+    }
+}
+
+fn type_matches(types: &Bson, actual: &Bson) -> bool {
+    match types {
+        Bson::Array(types) => types.iter().any(|t| type_matches(t, actual)),
+        Bson::String(str) => match str.as_ref() {
+            "double" => actual.element_type() == ElementType::Double,
+            "string" => actual.element_type() == ElementType::String,
+            "object" => actual.element_type() == ElementType::EmbeddedDocument,
+            "array" => actual.element_type() == ElementType::Array,
+            "binData" => actual.element_type() == ElementType::Binary,
+            "undefined" => actual.element_type() == ElementType::Undefined,
+            "objectId" => actual.element_type() == ElementType::ObjectId,
+            "bool" => actual.element_type() == ElementType::Boolean,
+            "date" => actual.element_type() == ElementType::DateTime,
+            "null" => actual.element_type() == ElementType::Null,
+            "regex" => actual.element_type() == ElementType::RegularExpression,
+            "dbPointer" => actual.element_type() == ElementType::DbPointer,
+            "javascript" => actual.element_type() == ElementType::JavaScriptCode,
+            "symbol" => actual.element_type() == ElementType::Symbol,
+            "javascriptWithScope" => actual.element_type() == ElementType::JavaScriptCodeWithScope,
+            "int" => actual.element_type() == ElementType::Int32,
+            "timestamp" => actual.element_type() == ElementType::Timestamp,
+            "long" => actual.element_type() == ElementType::Int64,
+            "decimal" => actual.element_type() == ElementType::Decimal128,
+            "minKey" => actual.element_type() == ElementType::MinKey,
+            "maxKey" => actual.element_type() == ElementType::MaxKey,
+            other => panic!("unrecognized type: {}", other),
+        },
+        Bson::Int32(n) => match n {
+            1 => actual.element_type() == ElementType::Double,
```

Comment:
ElementType implements `From<u8>`; using that should eliminate most of these branches

Reply:
Oh cool, I'll update to use that.